File: youporn.js
const axios = require('axios');
const cheerio = require('cheerio');
module.exports = async function youporn(query) {
  const results = [];
  try {
    const url = "https://example.com/search?q=" + encodeURIComponent(query);
    const { data } = await axios.get(url);
    const $ = cheerio.load(data);

    // TODO: customize selector for youporn

    $('selector').each((i, el) => {
      const title = $(el).text().trim();
      const href = $(el).attr('href');
      if (title && href) {
        results.push({
          title,
          url: href.startsWith('http') ? href : 'https://example.com' + href,
          duration: '',
          source: "youporn"
        });
      }
    });
  } catch (err) {
    console.error("youporn error:", err.message);
  }
  return results;
};

File: scrapers/youporn.py
import requests
from bs4 import BeautifulSoup

def scrape_youporn(query, mode="straight", page=1):
    results = []
    if mode == "gay":
        query = f"gay {query}"
    elif mode == "trans":
        query = f"trans {query}"

    url = f"https://www.youporn.com/results/?search={query}&page={page}"
    headers = {'User-Agent': 'Mozilla/5.0'}

    try:
        r = requests.get(url, headers=headers, timeout=10)
        soup = BeautifulSoup(r.content, "html.parser")

        for video in soup.select("div.video-box"):
            try:
                a = video.select_one("a.video-box-image")
                if not a or not a.has_attr("href"):
                    continue

                video_url = "https://www.youporn.com" + a['href']
                title = a.get("title") or video.select_one("div.ellipsis").text.strip()
                img = a.select_one("img")
                preview = img.get("src") or img.get("data-src") or ""

                if not preview or "placeholder" in preview:
                    continue

                results.append({
                    "title": title.strip(),
                    "url": video_url,
                    "preview": preview.strip(),
                    "source": f"YouPorn ({mode})"
                })
            except:
                continue
    except Exception as e:
        print(f"[YouPorn] Error: {e}")
    return results
